src/test/lib/DataTest.cpp

```diff
@@ -4509,4 +4509,199 @@ QuicTestStreamMultiReceive(
+        TEST_QUIC_SUCCEEDED(ClientStream.Send(&Buffer, 1));
+
+        ReceiveContext.StreamClosed.WaitTimeout(TestWaitTimeout);
+        TEST_EQUAL(ReceiveContext.ReceivedBytes, BufferSize);
+        TEST_EQUAL(0, memcmp(SendDataBuffer, ReceiveDataBuffer, BufferSize));
+    }
+
+    // Server side sending data
+    {
+        // Setup a listener
+        AppBuffersSenderContext SenderContext{};
+        MsQuicAutoAcceptListener Listener(Registration, ServerConfiguration, AppBuffersSenderContext::ConnCallback, &SenderContext);
+        TEST_QUIC_SUCCEEDED(Listener.GetInitStatus());
+        TEST_QUIC_SUCCEEDED(Listener.Start("MsQuicTest"));
+        QuicAddr ServerLocalAddr;
+        TEST_QUIC_SUCCEEDED(Listener.GetLocalAddr(ServerLocalAddr));
+
+        // Setup a client connection
+        MsQuicConnection Connection(Registration);
+        TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+
+        TEST_QUIC_SUCCEEDED(Connection.Start(
+            ClientConfiguration,
+            ServerLocalAddr.GetFamily(),
+            QUIC_TEST_LOOPBACK_FOR_AF(ServerLocalAddr.GetFamily()),
+            ServerLocalAddr.GetPort()));
+        TEST_TRUE(Connection.HandshakeCompleteEvent.WaitTimeout(TestWaitTimeout));
+        TEST_TRUE(Connection.HandshakeComplete);
+
+
+        // Create send and receive buffers
+        const uint32_t BufferSize = 0x5000;
+        uint8_t SendDataBuffer[BufferSize] = {};
+        for (int i = 0; i < BufferSize; ++i) {
+            SendDataBuffer[i] = static_cast<uint8_t>(i);
+        }
+
+        uint8_t ReceiveDataBuffer[BufferSize] = {};
+        QUIC_BUFFER QuicBuffers[5]{};
+        for (auto i = 0u; i < 5; ++i) {
+            QuicBuffers[i].Buffer = ReceiveDataBuffer + i * BufferSize / 5;
+            QuicBuffers[i].Length = BufferSize / 5;
+        }
+
+        // Create and start a stream
+        AppBuffersReceiverContext ReceiveContext;
+
+        MsQuicStream ClientStream(
+            Connection,
+            QUIC_STREAM_OPEN_FLAG_EXTERNAL_BUFFERS,
+            CleanUpManual,
+            AppBuffersReceiverContext::StreamCallback,
+            &ReceiveContext);
+
+        TEST_QUIC_SUCCEEDED(ClientStream.GetInitStatus());
+
+        // Provide receive buffers before starting the stream
+        MsQuic->StreamProvideReceiveBuffers(
+            ClientStream,
+            ARRAYSIZE(QuicBuffers),
+            QuicBuffers);
+
+        TEST_QUIC_SUCCEEDED(ClientStream.Start(QUIC_STREAM_START_FLAG_IMMEDIATE));
+
+        auto* SenderStream = SenderContext.WaitForSenderStream();
+
+        // Send data
+        QUIC_BUFFER Buffer{BufferSize, SendDataBuffer};
+        TEST_QUIC_SUCCEEDED(SenderStream->Send(&Buffer, 1));
+
+        ReceiveContext.StreamClosed.WaitTimeout(TestWaitTimeout);
+        TEST_EQUAL(ReceiveContext.ReceivedBytes, BufferSize);
+        TEST_EQUAL(0, memcmp(SendDataBuffer, ReceiveDataBuffer, BufferSize));
+    }
+}
 #endif // QUIC_API_ENABLE_PREVIEW_FEATURES
```

Comment:
Why does this need to be in the preview features section?

Reply:
I haven't done it yet, but I assumed the entire new API should go initially in the preview feature, until we treat it stable.